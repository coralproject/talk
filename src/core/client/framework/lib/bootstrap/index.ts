export * from "./TalkContext";
export { default as createContext } from "./createContext";
export { default as withContext } from "./withContext";
