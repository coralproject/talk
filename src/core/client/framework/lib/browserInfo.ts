import Bowser from "bowser";

export interface BrowserInfo {
  version: number;
  ios: boolean;
  msie: boolean;
  mobile: boolean;
  supports: {
    cssVariables: boolean;
    fetch: boolean;
    proxyObject: boolean;
    intl: boolean;
    intlPluralRules: boolean;
    intersectionObserver: boolean;
  };
}

let browserInfo: BrowserInfo | null = null;

export function getBrowserInfo(): BrowserInfo {
  if (!browserInfo) {
    const browser = Bowser.getParser(window.navigator.userAgent);
    const ios = browser.is(Bowser.OS_MAP.iOS);
    const msie = browser.is(Bowser.BROWSER_MAP.ie);
    const mobile = browser.is(Bowser.PLATFORMS_MAP.mobile);
    const version = Number.parseFloat(browser.getBrowserVersion());
    browserInfo = {
      version,
      supports: {
        intl: typeof Intl !== "undefined",
        intlPluralRules:
          typeof Intl !== "undefined" && Boolean(Intl.PluralRules),
        proxyObject: Boolean(window.Proxy),
        cssVariables: window.CSS && CSS.supports("color", "var(--fake-var)"),
        fetch: Boolean(window.fetch),
        intersectionObserver:
          "IntersectionObserver" in window &&
          "IntersectionObserverEntry" in window &&
          "intersectionRatio" in
            (window as any).IntersectionObserverEntry.prototype,
      },
      ios,
      msie,
      mobile,
    };
  }
  return browserInfo;
}
