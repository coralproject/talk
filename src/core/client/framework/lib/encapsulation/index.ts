export * from "coral-ui/encapsulation";
export { default as IframeEncapsulation } from "./IframeEncapsulation";
