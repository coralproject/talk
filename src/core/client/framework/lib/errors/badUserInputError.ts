import { mapValues, once } from "lodash";
import { ReactNode } from "react";
import { VALIDATION_REQUIRED } from "../messages";

/**
 * ValidationError represents all possible string values
 * that is responded by the server.
 */
type ValidationError = "REQUIRED";

/**
 * InvalidArgsMap as responded by the server.
 */
interface InvalidArgsMap {
  [key: string]: ValidationError;
}

/**
 * The localized version of `InvalidArgsMap`.
 */
interface InvalidArgsMapLocalilzed {
  [key: string]: ReactNode;
}

/**
 * Shape of the `BadUserInput` extension.
 */
interface BadUserInputExtension {
  code: "BAD_USER_INPUT";
  exception: {
    invalidArgs: InvalidArgsMap;
  };
}

/**
 * Map server `ValidationError` to a translation message.
 */
const validationMap = {
  REQUIRED: VALIDATION_REQUIRED,
};

/**
 * BadUserInputError wraps the `BAD_USER_INPUT` error returned from the
 * server.
 */
export default class BadUserInputError extends Error {
  // Keep origin of original server response.
  public readonly origin: BadUserInputExtension;

  constructor(error: BadUserInputExtension) {
    super("BadUserInputError");

    // Maintains proper stack trace for where our error was thrown.
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, BadUserInputError);
    }

    this.origin = error;
  }

  get invalidArgs(): InvalidArgsMap {
    return this.origin.exception.invalidArgs;
  }

  get invalidArgsLocalized(): InvalidArgsMapLocalilzed {
    return this.computeInvalidArgsLocalized();
  }

  // Perform localization and memoize result.
  private computeInvalidArgsLocalized = once(() => {
    return mapValues(this.invalidArgs, v => {
      if (v in validationMap) {
        return validationMap[v]();
      }
      return v;
    });
  });
}
