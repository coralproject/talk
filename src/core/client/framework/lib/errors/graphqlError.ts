export interface GraphQLErrorItem {
  message: string;
  locations: Array<{
    line: number;
    column: number;
  }>;
}

/**
 * Graphql wraps graphql errors at the network layer.
 */
export default class GraphQLError extends Error {
  // Original error.
  public readonly origin: GraphQLErrorItem[];

  constructor(origin: GraphQLErrorItem[]) {
    super(origin.map(o => o.message).join(" "));

    // Maintains proper stack trace for where our error was thrown.
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, GraphQLError);
    }
    this.origin = origin;
  }
}
