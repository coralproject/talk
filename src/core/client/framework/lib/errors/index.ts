export { default as UnknownServerError } from "./unknownServerError";
export { default as BadUserInputError } from "./badUserInputError";
