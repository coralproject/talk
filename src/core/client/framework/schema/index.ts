export * from "./__generated__/types";
export * from "./custom";
