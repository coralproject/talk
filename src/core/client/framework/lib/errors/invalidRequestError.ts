import { FORM_ERROR } from "final-form";

import { ERROR_CODES } from "coral-common/errors";

/**
 * Shape of the `InvalidRequest` extension as
 * the client requires. Note: the only crucial
 * field is the `code` field.
 */
interface InvalidRequestExtension {
  code: ERROR_CODES;
  message?: string;
  id?: string;
  param?: string;
}

/**
 * InvalidRequestError wraps the `INVALID_REQUEST_ERROR` error returned from the
 * server.
 */
export default class InvalidRequestError extends Error
  implements InvalidRequestExtension {
  // Keep extension of original server response.
  public readonly extension: InvalidRequestExtension;
  public readonly code: ERROR_CODES;
  public readonly id?: string;
  public readonly param?: string;
  public readonly message: string;
  public readonly extensions: string;

  constructor(extension: InvalidRequestExtension) {
    super("InvalidRequestError");

    // Maintains proper stack trace for where our error was thrown.
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, InvalidRequestError);
    }
    this.extension = extension;
    this.code = extension.code;
    this.id = extension.id;
    this.param = extension.param;
    this.message = extension.message || extension.code;
  }

  public get invalidArgs() {
    if (this.param) {
      return {
        [this.param.substr("input.".length)]: this.message,
      };
    }
    return {
      [FORM_ERROR]: this.message,
    };
  }
}
