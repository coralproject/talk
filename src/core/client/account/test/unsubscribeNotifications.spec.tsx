import sinon from "sinon";

import { ERROR_CODES } from "coral-common/errors";
import { InvalidRequestError } from "coral-framework/lib/errors";
import { GQLResolver } from "coral-framework/schema";
import {
  act,
  createAccessToken,
  CreateTestRendererParams,
  replaceHistoryLocation,
  waitForElement,
  within,
} from "coral-framework/testHelpers";

import create from "./create";

const token = createAccessToken();

async function createTestRenderer(
  params: CreateTestRendererParams<GQLResolver> = {}
) {
  const { testRenderer, context } = create();
  return {
    context,
    testRenderer,
    root: testRenderer.root,
  };
}

it("renders missing confirm token", async () => {
  replaceHistoryLocation("http://localhost/account/notifications/unsubscribe");
  const { root } = await createTestRenderer();
  await waitForElement(() => within(root).getByTestID("invalid-link"));
  expect(within(root).toJSON()).toMatchSnapshot();
});

it("renders form", async () => {
  replaceHistoryLocation(
    `http://localhost/account/notifications/unsubscribe#unsubscribeToken=${token}`
  );
  const { root, context } = await createTestRenderer();

  const restMock = sinon.mock(context.rest);
  restMock
    .expects("fetch")
    .withArgs("/account/notifications/unsubscribe", {
      method: "GET",
      token,
    })
    .once();

  await act(async () => {
    await waitForElement(() => within(root).getByTestID("unsubscribe-form"));
  });
  expect(within(root).toJSON()).toMatchSnapshot();
  restMock.verify();
});

it("renders error from server", async () => {
  replaceHistoryLocation(
    `http://localhost/account/notifications/unsubscribe#unsubscribeToken=${token}`
  );

  const codes = [
    ERROR_CODES.RATE_LIMIT_EXCEEDED,
    ERROR_CODES.INTEGRATION_DISABLED,
    ERROR_CODES.USER_NOT_FOUND,
    ERROR_CODES.TOKEN_INVALID,
  ];

  for (const code of codes) {
    const { root, context } = await createTestRenderer();

    const restMock = sinon.mock(context.rest);
    restMock
      .expects("fetch")
      .withArgs("/account/notifications/unsubscribe", {
        method: "GET",
        token,
      })
      .once()
      .throwsException(
        new InvalidRequestError({
          code,
        })
      );

    await act(async () => {
      await waitForElement(() =>
        within(root).getByText(code, {
          exact: false,
        })
      );
    });
    restMock.verify();
  }
});

it("submits form", async () => {
  replaceHistoryLocation(
    `http://localhost/account/notifications/unsubscribe#unsubscribeToken=${token}`
  );
  const { root, context } = await createTestRenderer();

  const restMock = sinon.mock(context.rest);
  restMock
    .expects("fetch")
    .withArgs("/account/notifications/unsubscribe", {
      method: "GET",
      token,
    })
    .once();

  restMock
    .expects("fetch")
    .withArgs("/account/notifications/unsubscribe", {
      method: "DELETE",
      token,
    })
    .once();

  await act(async () => {
    await waitForElement(() => within(root).getByTestID("unsubscribe-form"));
  });
  const form = within(root).getByType("form");

  // Submit valid form.
  await act(async () => {
    form.props.onSubmit();
    await waitForElement(() => within(root).getByTestID("success"));
  });

  restMock.verify();
});
