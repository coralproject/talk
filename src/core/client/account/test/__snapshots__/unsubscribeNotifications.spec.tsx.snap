// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form 1`] = `
<div
  data-testid="main-layout"
>
  <div
    className="MainLayout-bar"
  />
  <div
    className="MainLayout-centered"
  >
    <div
      className="UnsubscribeRoute-container"
    >
      <div
        className="UnsubscribeRoute-root"
      >
        <div
          data-testid="unsubscribe-form"
        >
          <form
            onSubmit={[Function]}
          >
            <div
              className="Box-root HorizontalGutter-root HorizontalGutter-full"
            >
              <h1
                className="Box-root Typography-root Typography-heading1 Typography-colorTextPrimary"
              >
                Click below to confirm that you want to unsubscribe from all notifications.
              </h1>
              <button
                className="BaseButton-root Button-root Button-sizeRegular Button-colorPrimary Button-variantFilled Button-fullWidth"
                data-color="primary"
                data-variant="filled"
                disabled={false}
                onBlur={[Function]}
                onFocus={[Function]}
                onMouseOut={[Function]}
                onMouseOver={[Function]}
                onTouchEnd={[Function]}
                type="submit"
              >
                Confirm
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders missing confirm token 1`] = `
<div
  data-testid="main-layout"
>
  <div
    className="MainLayout-bar"
  />
  <div
    className="MainLayout-centered"
  >
    <div
      className="UnsubscribeRoute-container"
    >
      <div
        className="UnsubscribeRoute-root"
      >
        <div
          className="Box-root HorizontalGutter-root HorizontalGutter-double"
        >
          <h1
            className="Box-root Typography-root Typography-heading1 Typography-colorTextPrimary"
          >
            Oops Sorry!
          </h1>
          <div
            className="CallOut-root CallOut-colorError CallOut-fullWidth"
          >
            <div>
              <span
                data-testid="invalid-link"
              >
                The specified link is invalid, check to see if it was copied correctly.
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
